### DISCUSSION TITLE
How to export a custom module then import to another site?

### DISCUSSION THREAD

**Discussion by orchardc** [OrchardCMS/OrchardCore#12397] [Q&A] [answered]
I must miss something here... Previously there was an Install Local Module function; I know it has been replaced with the 'Deployment Plan', but I couldn't find a way to export my custom module (a feature), then import it to another Orchard site.

Is there any documentation for that deployment please?

**Comment by Skrypt** [comment, marked as answer]
You just need to enable that feature. But you can upload the entire module, it needs to be there first.

**Reply by orchardc** [reply]
Yes, but that's the part I'm struggeled with: "upload the entire module", how to do that?

**Reply by Skrypt** [reply]
You can't. You need to publish the website on the production and/or staging server first.
You can also use Docker for that matter to switch between containers "live". We don't provide a way to upload modules and build them on runtime. It was a bad design as it was causing the tenants to be unavailable while it was doing so.

Orchard Core is not Wordpress.

**Reply by orchardc** [reply]
OK, got you thanks. I'm thinking the following scenario: 

Imagine using the multi-tenant feature to host multiple websites, and each website has its own custom module.

Is it the only way to do it to add them as individual projects dependience on the same solution - so that custom modules can appear in the feature list?

How about maintenance for a custom module (on one site) ? I expect each project (site) can be deployed separately, without impacting others. Now it seems not possible to do that, and we can only deploy the whole solution (that includes multiple projects) all together?

**Reply by Skrypt** [reply]
You can keep different versions of one module for each of your tenants if that is your goal but then they need to be 2 different features. So, you would keep OldAssembly for Tenant1 and NewAssembly for Tenant2. Because that is what you ultimately want to do. Just like if we had OrchardCore.Workflows and OrchardCore.ElsaWorkflow which are 2 different implementations. Of course, that makes things harder to maintain over time, but normally, if you want to avoid this then you migrate the tenants that are using the old module and their integrations. It is all up to you at that point. 😄

**Reply by orchardc** [reply]
Thanks but sorry I think I didn't make it clear enough :)

Let's say we have two tenants and they have different customization:
  Tenant A - with custom module A
  Tenant B - with custom module B

I assume the only way to do that in Visual Studio is to include both of them into the same Visual Studio solution, like this: 
 - 1 CMS project  (dotnet new occms), has project references of the following two:
 - 1 module A project (dotnet new ocmodulecms)
 - 1 module B project (dotnet new ocmodulecms)

As the result, both modules will show in tenants, but I will only enable module-A for tenant-A, and module-B for tenant-B.

All good so far... until...If I need to change something in module-A. I will have to make the change in module-A project and then publish the whole solution (CMS, module A and B), correct? This is not ideal, because it would be much better to just make the change in the module-A project, and then export and import the module into Production. Even if that approach is not possible, at least there should be a way to just publish module-A, rather than everything. Is it possible and how?

I hope it's clear now.

**Reply by Skrypt** [reply]
There is a feature in Orchard to make a module available only for a specific tenant. That's one thing.

The other one about updating a module is that you have a Migrations.cs file that is available for you to run any kind of migrations on your custom modules. So, you can run a recipe from these migrations if needed. So, yes, when you update a module you need to update it manually and it will run these migrations for you based on which migrations step has run before or not.

Also, it is possible to publish only the module that you want to publish and update its assembly on the server without requiring to update every assembly if they are using the same .NET framework and Orchard Core version. If there is an update of .NET framework between module updates then it requires also that you update every other assembly that your module uses.

That's one of the reasons why uploading a module is not that great because there are also module dependencies that often require to be updated. And in Orchard 1 we had issues with that where people were trying to install modules that were not working anymore because they required to be updated for the newer versions of Orchard.

Again, this is not a PHP world where you can install/uninstall PHP code by removing the code from the server. These are assemblies that have running instances on the server which need to be disposed of before installing an updated assembly. And there is no such thing as removing an assembly from runtime. Also adding an assembly on runtime requires getting the graph of all its dependencies which is quite a pain when those are updated by other custom modules.

In the end, the better solution is to publish module updates manually to a staging server first and then push them to the prod server once it is tested. This can be leveraged through a CI or by using Docker containers locally or on a prod server. That's the advised way to do things now.

**Reply by orchardc** [reply]
Thanks Skrypt, yes I get it. 

Could you clarify one more thing, when you say _"Also, it is possible to publish only the module that you want to publish and update its assembly on the server without requiring to update every assembly if they are using the same .NET framework and Orchard Core version. "_ 

See the screenshot below, I can right-click the MultiTenantApplication, and click Publish, then publish everything to Azure.
How can I just publish Module1 to Azure without impacting others? They are all in OC1.4 with .Net6.

![image](https://user-images.githubusercontent.com/16193516/189725056-664d39b4-12a6-455f-982e-3901c6ff0c67.png)


This will solve my problem and indeed there are staging servers between DEV and Prod.

Jim

**Reply by Skrypt** [reply]
Publish to a local folder. Upload through FTP the required updated assembly.

**Reply by orchardc** [reply]
Ok got it, thanks!
